net.solarnetwork.node.setup.web/web/js/graph-data.js

```diff
@@ -0,0 +1,237 @@
+function onlyUnique(value, index, self) { 
+    return self.indexOf(value) === index;
+}
+
+//https://stackoverflow.com/questions/7343890/standard-deviation-javascript
+//This get used for giving the graph some white space either side of the graph
+function StandardDeviation(numbersArr) {
+    //--CALCULATE AVAREGE--
+    var total = 0;
+    for(var key in numbersArr) 
+       total += numbersArr[key];
+    var meanVal = total / numbersArr.length;
+    //--CALCULATE AVAREGE--
+  
+    //--CALCULATE STANDARD DEVIATION--
+    var SDprep = 0;
+    for(var key in numbersArr) 
+       SDprep += Math.pow((parseFloat(numbersArr[key]) - meanVal),2);
+    var SDresult = Math.sqrt(SDprep/numbersArr.length);
+    //--CALCULATE STANDARD DEVIATION--
+    return SDresult
+    
+}
+
+//when a new datum comes in, this handler gets called
+//the handler checks if the new datum has a wattage reading and 
+var handler = function handleMessage(msg) {
+	
+	var datum = JSON.parse(msg.body).data;
+
+	//check that the datum has a reading
+	if (datum[key] != undefined){
+		//map the sourceId to the current reading
+		datamap[datum.sourceId] = datum[key];
+		
+		//if we have not seen this sourceId before we need to graph it
+		if (sourceIds.indexOf(datum.sourceId)==-1){
+			
+			//add a new graph for this new sourceId
+			sourceIds.push(datum.sourceId);
+			graphinit(datum.sourceId,units);
+		}
+	}
+};
+
+//creates a new graph looking that plots wattage data from datums coming from source 
+//this code is heavily based on the last graph from https://bost.ocks.org/mike/path/
+//apologies for the magic numbers 
+function graphinit(source, units) {
+
+
+    var n = 243,//how many sample points to have on the graph
+        duration = 1000,//time for the animation 
+        now = new Date(Date.now() - duration),//not sure what the -duration is for
+
+        //prefill the array with the first reading
+        data = new Array(n).fill(datamap[source]);
+
+    //positional styling for the graph
+    var margin = { top: 10, right: 0, bottom: 20, left: 60 },
+        width = 960 - margin.right,
+        height = 120 - margin.top - margin.bottom;
+
+    //sets the axis scales
+    var x = d3.time.scale()
+        .domain([now - (n - 2) * duration, now - duration])
+        .range([0, width]);
+
+    var y = d3.scale.linear()
+        .range([height, 0]);
+
+    //draws the line for the graph (not sure how this code works at this stage)
+    var line = d3.svg.line()
+        .interpolate("step-after")//other possible options are linear, step-after, basis, bundle and cardinal
+        .x(function (d, i) { return x(now - (n - 1 - i) * duration); })//not sure what is going on here 
+        .y(function (d, i) { return y(d); });//not sure what is going on here
+
+    //finds the location on the page the script is loaded is loaded to put the graph
+    var p = d3.select(".graphpoint").append("p").text(source + " (" + key + ")");//adds a title in form of "SourceId (metric)"
```

Comment:
A `<h3>` element would be more appropriate than `<p>` here I think for chart titles.